### PR TITLE
Instantiate NMSInterface implementations dynamically with Reflection

### PR DESCRIPTION
**Pull request by Sytm**
- Instantiate NMSInterface implementations dynamically with Reflection, simplifies code for selecting implementation and adding new ones
- Fixed NMSInterface for Minecraft 1.15 to actually work on Minecraft 1.15
- Added NMSInterface implementation for Paper servers without actual NMS code that works independent of server version

The only drawback to using Reflection like it is implemented at the moment is that developers need to exclude pathetic from minimizing or else the classes will be removed. (Although this can be changed if required by you)
---
Download the Example Plugin jar for this pull request: [pathetic-25.zip](https://nightly.link/patheloper/pathetic/actions/artifacts/722508019.zip)

### CODE REVIEW

File: pathetic-nms/src/main/java/org/patheloper/nms/NMSUtils.java [hunk 1]
```diff
@@ -1,56 +1,43 @@
...
+                        (NMSInterface) Class.forName(getClass().getPackage().getName() + ".NMSInterface_" + nmsVersion)
+                                .getConstructor()
+                                .newInstance();
+            } catch (ClassNotFoundException e) {
+                throw new IllegalArgumentException("Unsupported version: " + nmsVersion, e);
+            } catch (ReflectiveOperationException e) {
+                throw new IllegalArgumentException("Could not create NMSInterface for version " + nmsVersion, e);
+            }
         }
+        System.out.println("Used NMS interface: " + nmsInterface);
```
Comment (olijeffers0n) [inline pathetic-nms/src/main/java/org/patheloper/nms/NMSUtils.java]: Debug Message that got missed?

File: pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterface_Paper.java [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+public class NMSInterface_Paper implements NMSInterface {
+
+    @Override
+    public ChunkSnapshot getSnapshot(World world, int chunkX, int chunkZ) {
+        if (Bukkit.isPrimaryThread()) {
+            // If we are currently on the primary thread and call getChunkAtAsync on paper
+            // we would enter a deadlock when using .join()
+            return world.getChunkAt(chunkX, chunkZ).getChunkSnapshot();
+        } else {
+            return world.getChunkAtAsync(chunkX, chunkZ).join().getChunkSnapshot();
```
Comment (olijeffers0n) [inline pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterface_Paper.java]: Maybe look at [`getChunkAtAsyncUrgently`](https://jd.papermc.io/paper/1.19/org/bukkit/World.html#getChunkAtAsyncUrgently(int,int))

File: pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterface_Paper.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package org.patheloper.nms;
+
+import org.bukkit.Bukkit;
+import org.bukkit.ChunkSnapshot;
+import org.bukkit.World;
+import org.patheloper.api.snapshot.NMSInterface;
+
+public class NMSInterface_Paper implements NMSInterface {
```
Comment (olijeffers0n) [inline pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterface_Paper.java]: If we are trying to stick to proper naming conventions there shouldn't be Underscores in class names

**Review by olijeffers0n — CHANGES_REQUESTED** [review]
Assuming the reflective class name logic works, I think this looks good. Just a few things to address though. Thanks!

**Review by olijeffers0n — APPROVED** [review]
Approved subject to @GodCipher 's approval and testing

**Comment by bsommerfeld** [comment]
first of all, hi and thanks for contributing.

secondly:
the bugfix for the 1.15 is very useful and we will take that into account, thanks.

but other than that:
please revert the renaming of the NMS modules. we prefer to maintain a more explicit and controlled method to ensure stability and avoid potential issues or drawbacks. 

additionally, i notcied you added extra support for paper, which we have [previously agreed](https://github.com/patheloper/pathetic/issues/21#issuecomment-1567485377) not to include in our project. while it may provide some benefits, we have decided to avoid introducing dependencies that could potentially complicate our codebase and create unnecessary maintenance overhead.

so if there is not an important reason to keep this, i kindly ask you to revert that change as well.

for the future, please remember to maintain some granularity so the main point of the PR doesnt get out of hand.

File: pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java
Comment (bsommerfeld) [inline pathetic-model/src/main/java/org/patheloper/model/snapshot/FailingSnapshotManager.java:1]: as mentioned, revert this please.

File: pathetic-nms/paper/pom.xml
Comment (bsommerfeld) [inline pathetic-nms/paper/pom.xml:1]: ^same for this one

File: pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterfacePaper.java
Comment (bsommerfeld) [inline pathetic-nms/paper/src/main/java/org/patheloper/nms/NMSInterfacePaper.java:1]: +1

File: pathetic-nms/pom.xml
Comment (bsommerfeld) [inline pathetic-nms/pom.xml:1]: +1

File: pathetic-nms/src/main/java/org/patheloper/nms/NMSUtils.java
Comment (bsommerfeld) [inline pathetic-nms/src/main/java/org/patheloper/nms/NMSUtils.java:1]: +1 just like the following renamings

File: pom.xml
Comment (bsommerfeld) [inline pom.xml:1]: +1

File: pathetic-nms/v1_15/src/main/java/org/patheloper/nms/NMSInterfacev1_15_R1.java
Comment (bsommerfeld) [inline pathetic-nms/v1_15/src/main/java/org/patheloper/nms/NMSInterfacev1_15_R1.java:1]: the fix is great, but revert the renaming

File: pathetic-nms/v1_15/pom.xml [hunk 4]
```diff
@@ -29,7 +29,7 @@
         <dependency>
             <groupId>org.spigotmc</groupId>
             <artifactId>spigot</artifactId>
-            <version>1.16.5-R0.1-SNAPSHOT</version>
+            <version>1.15.2-R0.1-SNAPSHOT</version>
```
Comment (bsommerfeld) [inline pathetic-nms/v1_15/pom.xml:32]: good catch!

**Review by bsommerfeld — CHANGES_REQUESTED** [review]

**Comment by Sytm** [comment]
I wouldn't say that the Paper hook complicates the codebase, especially compared to the undocumented NMS shenanigans on newer versions. And Paper isn't just _some_ fork of Spigot, but used more frequently by server owners than pure Spigot and this API allows cleaner handling of chunks.

Maintenance overhead is basically zero because this does not require updating to newer versions, but only needs to be added once (compared to NMS), so I thought I could do that for you.

But if that is still not wanted I will remove it later today while addressing the other changes.

**Comment by bsommerfeld** [comment]
it adds another dependency which we would have to maintain. despite that it would not fit into the NMS module. means that it would need some overwork too.

with every update of paper we would need to verify that this hook still works like intended while the other NMS impls stay the same.

> But if that is still not wanted I will remove it later today while addressing the other changes

that would be great

**Comment by bsommerfeld** [comment]
there are still some changes left before we can merge that into pathetic

**Comment by Sytm** [comment]
Which changes? I have changed it according to your feedback with the original purpose of this PR in mind.

**Comment by bsommerfeld** [comment]
there are still requested changes open

**Comment by Sytm** [comment]
_Which changes exactly are still required?_

**Comment by olijeffers0n** [comment]
Firstly, on behalf of the contributors of Pathetic, I would like to say thank you for the work you put into this PR. 

But alas, let us now delve into the captivating topic of the changes you've made. Oh, what a delightful surprise it is to discover that these changes bear absolutely no resemblance to the glorious path we, as a project, have so meticulously paved. And might I add, your audacity to shamelessly clone our work and masquerade it as your very own creation is simply awe-inspiring. Truly, we are left in utter admiration of your remarkable display of bad faith. Kudos!

In light of these earth-shattering revelations, it brings us immense pleasure to inform you that we have made the executive decision not to move forward with these extraordinary changes. Yes, you heard it right! We are exercising our magnanimous power to put a halt to your valiant efforts in their tracks. How fortunate we are to possess such wisdom and discernment!

Fear not, my dear copy-paste aficionado, for we shall not dismiss your contribution entirely. Oh no, we are far too kind for that. Instead, we shall graciously pull your bug changes into a separate commit. Such benevolence! Please, do accept our heartfelt appreciation for stumbling upon these bugs. We are eternally indebted to your unparalleled bug-finding prowess.

In conclusion, I sincerely wish you the absolute best in all your future copy-paste endeavours. May you continue to spread the joy of unoriginality and reap the rewards of your extraordinary talent. Bravo, good sir, bravo!
